### PR TITLE
1.0

### LINKED ISSUE #101: Release 1.0?
This crate has become a pretty foundational part of a few Rust ecosystems:

- Embedded (via `heapless` and other fixed-sized containers)
- RustCrypto (where we just completed a `generic-array` v0.14 upgrade)

Unfortunately, if you now use both `heapless` and RustCrypto in conjunction (via RustCrypto's `aead::heapless` for example) it pulls in `generic-array` v0.12, v0.13, and v0.14! (this is due to the [`as-slice`](https://github.com/japaric/as-slice/blob/master/Cargo.toml#L13-L14) crate, which pulls in both v0.12 and v0.13 for whatever reason)

In addition to eliminating confusion around coordinating `generic-array` upgrades, a 1.0 release of `generic-array` is one of the main blockers for a 1.0 release of many of the RustCrypto crates, so it'd be fantastic if that were to happen.

Any thoughts on a 1.0 release?

### LINKED ISSUE #107: Make `'static` part of `ArrayLength` trait bounds?
In some recent code, for example here, https://docs.rs/generic-array/0.14.3/src/generic_array/sequence.rs.html#305, we have `+ 'static` bounds on types that represent `ArrayLength` typenums.

```
unsafe impl<'a, T, N, K> Split<T, K> for &'a GenericArray<T, N>
where
    N: ArrayLength<T>,
    K: ArrayLength<T> + 'static,
    N: Sub<K>,
    Diff<N, K>: ArrayLength<T>,
{
    type First = &'a GenericArray<T, K>;
    type Second = &'a GenericArray<T, Diff<N, K>>;

    fn split(self) -> (Self::First, Self::Second) {
        unsafe {
            let ptr_to_first: *const T = self.as_ptr();
            let head = &*(ptr_to_first as *const _);
            let tail = &*(ptr_to_first.add(K::USIZE) as *const _);
            (head, tail)
        }
    }
}
```

Sometimes in my own code that uses generic array, I have to put `'static` on the number type bounds, and it's sometimes hard to understand why.

It seems to me that we could change `ArrayLength` to require `'static`

```
pub unsafe trait ArrayLength<T>: Unsigned + 'static {
    /// Associated type representing the array type for the number
    type ArrayType;
}
```

Since this trait is only implemented for `typenum` types and those are all ZWT's anyway, the `'static` bound is always true and this breaks nothing in practice. It only helps the compiler to see that some things that are generic parameters are `'static` without the user having to spell it out.

Then you could remove `+ 'static` in several places in generic array itself, e.g.

```
    K: ArrayLength<T> + 'static,
```

in code like `impl ... Split` because the `+ 'static` would be implied.

What do you think?

### LINKED ISSUE #109: Broken links in doc.rs documentation
There are a few broken links in the documentation located at https://docs.rs/generic-array/0.14.4/generic_array/index.html. Namely, both `unsigned integer types` and `typenum` links are dead. The version build directly from the GitHub repo (located at https://fizyk20.github.io/generic-array/generic_array/) works though.

### LINKED ISSUE #119: support try_form?
GenericArray::from_slice(&[u8) will panic if slice length is not match. It would be great to provide a Result version, e.g. (GenericArray::try_from_slice(&[u8) -> Result<Self, LengthNotMatch>)

### LINKED ISSUE #132: Is there a way to convert GenericArray<T, N> **into** [T; N] ?

### LINKED ISSUE #135: Is the reason for this crate still valid with const generics? If so, maybe update docs.
This is the motivating example for this crate (slight difference between README and docs):
```
struct Foo<T, N> {
    data: [T; N]
}
```

Could this not be written (in recent Rust versions) as:
```
struct Foo<T, const N: usize> {
    data: [T; N]
}
```

If not, maybe update the motivating example?

### LINKED ISSUE #137: Simplify ArrayLength<T> trait to ArrayLength (make the ArrayLength::ArrayType GAT generic instead)
GenericArray<T, N> requires that `N` implements `ArrayLength<T>` which is currently implemented [like this](https://github.com/fizyk20/generic-array/blob/master/src/lib.rs#L116-L119):

```rust
pub unsafe trait ArrayLength<T>: Unsigned {
    /// Associated type representing the array type for the number
    type ArrayType;
}
```

However, , this is overly restrictive.
If you're working on any kind of trait which requires changing the GenericArray's element types, you end up with bounds like `N: ArrayLength<T> + ArrayLength<S> + ArrayLength<R> + ...`.

But `T` is only used at the place where the ArrayType associated type is actually inspected, and otherwise completely unconstrainted.

As such, in modern Rust versions (I'm not sure what the MSRV is), we could change the trait to:

```rust
pub unsafe trait ArrayLength: Unsigned {
    /// Associated type representing the array type for the number
    type ArrayType<T>;
}
```

which captures the intent more clearly, and keeps bounds on `N` much simpler.

### LINKED ISSUE #140: More generic From<[T; N]> and Into<[T; N]> implementations
Currently, these implementations are hard-coded for certain array lengths.

This was necessary on old Rust versions in which 
- const generics were not possible, and
- array traits were restricted to max lengths of 32. 
But since v1.47 [this is no longer the case](https://caniuse.rs/features/array_gt_32_impls).

By enabling the `consts` feature on `typenum`, we [get access to the `U<const N: usize>` type](https://github.com/paholg/typenum/pull/176). And this is what lets us write a single implementation for From and for Into that will work for all lengths!

### LINKED ISSUE #145: Feature request: Conversion from &[[T; N]] to &[GenericArray<T, N>] (and &mut)
APIs like the cipher crate's [`BlockEncrypt::encrypt_blocks`](https://docs.rs/cipher/latest/cipher/trait.BlockEncrypt.html#method.encrypt_blocks) take a slice of GenericArrays. However, if you're starting with a flat `&[u8]`, it's unergonomic to get to the required type. The unstable [`slice::as_chunks`](https://doc.rust-lang.org/std/primitive.slice.html#method.array_chunks) gives you `&[[u8; N]]`, but going from that to `&[GenericArray<u8, _>]` seems impossible without a pointer cast or transmute. However, that pointer cast or transmute *is* valid because GenericArray is `repr(transparent)`, and it would be nice™ if the generic-array crate provided that operation as a safe wrapper.

### PR DESCRIPTION
**Pull request by novacrazy**
With #137, this opens up the opportunity for a full 1.0 release with much-improved ergonomics. 

[CHANGELOG.md](https://github.com/fizyk20/generic-array/blob/1.0/CHANGELOG.md)

What other changes should we make while we can? 

Closes #145 
Closes #140 
Closes #137 
Closes #135 
Closes #132
Closes #119 
Closes #117 
Closes #109 
Closes #107 
Closes #101 

Additional tweaks:
* Hex encoding uses a single buffer for `N <= 1024` rather than `N < 1024`
* Hex encoding optimizations to avoid bounds checks and remove code duplication
* Tweaked where actual as_slice/deref/as_ref casts occur, with more inlining
* Simplified doc comments (removed `fn main()`, etc.)
* Adjusts readme to reflect modern Rust with `min_const_generics`
* Added an additional check inside the private const-`transmute` function to verify type sizes.
* Made the `ArrayType` implementations Sealed + unreachable Clone methods
* Added pre-iteration checks for `FromIterator` and deserialization routines
* Re-Optimized specialized zip routines for non-Drop items
* Added const variants of fallible APIs (`try_from_slice`)
* Added `GenericArray::len()` associated method without `&self`

Remaining todos:
- [x] ~~Update DESIGN.md~~ Moved into main docs.
- [x] Get #139 merged and setup, @fizyk20

### CODE REVIEW

**Comment by Qqwy** [comment]
These changes are great!

I have pointed my local project to the `1.0` branch and many trait implementations now become a lot simpler.


> What other changes should we make while we can?

Would it be possible to extend the `arr!` macro to also except a form without a type at the start?

Currently it supports only `let foo = arr![usize; 1, 2, 3]` and errors on e.g. `let foo = arr![1, 2, 3]`.
I think that in many situations, the explicit type is not neccesary because it might be inferred from the surroundings regardless.
Furthermore, it seems that currently `as` is used on elements not already matching the type rather than `Into::into()`, which breaks for any non-primitive types.

Let me know if this is the right place for such a request, or if you want me to create a separate issue for this.

**Comment by tarcieri** [comment]
#133 would be very helpful for several of us

**Comment by novacrazy** [comment]
> Would it be possible to extend the `arr!` macro to also accept a form without a type at the start?

Yes, quite easily. I'm considering making that the default, as `arr![]` is a tad annoying with the type and using `value as T` internally. Would probably be better to mirror `vec![]` in that regard, and simply force users to specify the type when assigning to a variable, so:

```rust
let value = arr![1u8, 2, 3, 4]; // GenericArray<u8, U4>
let value = arr![1, 2, 3, 4]; // GenericArray<i32, U4>, because i32 is the default integer type
let value = arr![1f32; U16]; // GenericArray<f32, U16>
let value: GenericArray<usize, _> = arr![4, 5, 6, 7]; // GenericArray<usize, U4>
```

**Comment by novacrazy** [comment]
I have changed the `arr![]` syntax. It is now more simplified and predictable. Changing 98% of the examples and tests required nothing more than removing the previous explicit type syntax or giving it the explicit type with the values themselves (e.g. `1u8`), similarly to how `vec![]` would work. No more internal `as` casts. Inference just works.

If there is opposition to this change, I can always revert it, but I feel like it's an improvement.

**Comment by novacrazy** [comment]
I've been experimenting with implementing `GenericSequence` and `FunctionalSequence` for `Box<GenericArray<T, N>>` as a way to manipulate heap-allocated `generic-array`s. It kind of works, but ends up incompatible with stack-allocated arrays, so `heap_array.zip(stack_array, ...)` doesn't work. That may be acceptable, though.

It does turn out I can cheat and use a `Vec<T>` (with fixed capacity) to build up a `Box<GenericArray<T, N>>` on the heap, then doing some pointer casting. No need for a heap-specific `ArrayBuilder`.

**Comment by novacrazy** [comment]
I've published `generic-array 1.0.0-alpha.0` for docs testing. Pre-release versions should never be automatically selected by Cargo, so it's pretty safe. I'll yank it later when 1.0 is fully published.

**Comment by novacrazy** [comment]
Seems that `doc_auto_cfg` isn't working on docs.rs. Perhaps I misunderstood how it works.

**Comment by Qqwy** [comment]
Thank you for publishing a prerelease version of the docs.

I am building a matrix multiplication library against the in-progress 1.0 branch and the improvements work very well.

One small suggestion for the docs:

> The [ArrayLength](https://docs.rs/generic-array/1.0.0-alpha.1/generic_array/trait.ArrayLength.html) trait is implemented by default for [unsigned integer types](https://docs.rs/typenum/1.16.0/x86_64-unknown-linux-gnu/typenum/marker_traits/trait.Unsigned.html) from [typenum](https://docs.rs/typenum/1.16.0/x86_64-unknown-linux-gnu/typenum/index.html):

Seems like since it is subtrait of `Unsigned`, the 'by default' part should probably be removed from that sentence.

---

The information about when people should and shouldn't need to mess with the `ArrayLength::ArrayType<T>` bound directly is a very good addition I think 👍 .

**Review by fizyk20 — APPROVED** [review]
This is awesome - I love the changes you made here. Thank you for the hard work you put into this!

**Comment by novacrazy** [comment]
```rust
macro_rules! arr {
    ($($x:expr),* $(,)*) => ( $crate::GenericArray::from_array([$($x),*]) );
    ($x:expr; $N:ty)     => ( $crate::GenericArray::from_array([$x; <$N as $crate::typenum::Unsigned>::USIZE]) );
    ($x:expr; $n:expr)   => ( $crate::GenericArray::from_array([$x; $n]) );
}
```
Turns out we can do this now. Massively simplified.

**Comment by novacrazy** [comment]
I've published `1.0.0-beta.2` on crates.io until this PR can be reviewed again. As of writing this, it's 125th in the docs.rs queue. I'll update with a link when it finishes.

**Comment by novacrazy** [comment]
After idly looking through the docs, I've decided to remove `TryFromIterError`, opting instead to use the single `LengthError`.

**Comment by novacrazy** [comment]
Would anyone else subscribed to this PR care to look over the code? More eyes the better.

**Comment by novacrazy** [comment]
I thought I'd fixed this years ago, but as it was there were some obvious autovectorization misses with `.zip()` especially. 

For example, before my latest commits, this code:
```rust
pub fn add64(x: GenericArray<f32, U64>, y: GenericArray<f32, U64>) -> GenericArray<f32, U64> {
    x.zip(y, |a, b| a + b)
}
```
would result in **64 serial `vaddss` instructions**, with 2 moves for each, which was absurd. 

Now it correctly autovectorizes to 8 `vaddps` AVX instructions, even without `codegen-units = 1` in my tests.

**Review by fizyk20 — APPROVED** [review]
Looks good to me in general, just a few small comments. Thanks again for all your hard work on this crate!

File: src/iter.rs [hunk 1]
```diff
@@ -155,19 +133,19 @@ where
             })
         };
 
-        // ensure the drop happens here after iteration
-        drop(self);
+        // ensure this happens here after iteration
+        mem::forget(self);
```
[the remark this answers is not in the document]
Comment (novacrazy) [reply src/iter.rs:137]: The iterator will be empty after folding the remaining items, everything moved out, so forgetting it simplifies codegen slightly by not inserting the drop call/code for the empty iterator. Drop is still called on panics, though, invisibly. I’ll add a comment about this.

File: src/arr.rs [hunk 2]
```diff
@@ -1,127 +1,96 @@
...
-        $crate::arr_impl!($T; $($x),*)
-    );
-    ($T:ty; $x:expr; $N:ty) => (
-        $crate::arr_impl!($T; $x; $N)
-    );
-    ($($x:expr,)+) => (arr![$($x),+]);
-    () => ("""Macro requires a type, e.g. `let array = arr![u32; 1, 2, 3];`")
+macro_rules! box_arr_helper {
+    (@count_ty) => { $crate::typenum::U0 };
+    (@count_ty $val:expr$(, $vals:expr)* $(,)?) => { $crate::typenum::Add1<$crate::box_arr_helper!(@count_ty $($vals),*)> };
```
[the remark this answers is not in the document]
Comment (novacrazy) [reply src/arr.rs]: I've solved this by deducing the length from an array of unit types constructed alongside the Vec, and using the same trick as `from_array` using typenum's `Const`, so now it supports the same range as regular `arr!`. The helper macro that does nothing but replace an expression with a unit type is a little annoying, but a better solution could be added eventually.

**Comment by novacrazy** [comment]
[`generic-array` 1.0.0 has been published!](https://crates.io/crates/generic-array/1.0.0)

**Pull request merged**